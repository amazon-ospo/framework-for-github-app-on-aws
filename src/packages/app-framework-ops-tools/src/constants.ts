import { name, version } from '../package.json';
// Tags for DynamoDB App Table
export enum TagName {
  FRAMEWORK_FOR_GITHUB_APP_ON_AWS_MANAGED = 'FrameworkForGitHubAppOnAwsManaged',
  CREDENTIAL_MANAGER = 'CredentialManager',
  APP_TABLE = 'AppTable',
}

// Key Specifications
// Docs: https://docs.aws.amazon.com/kms/latest/APIReference/API_GetParametersForImport.html#API_GetParametersForImport_RequestSyntax
export const WRAPPING_SPEC = 'RSA_4096';
// Docs: https://docs.aws.amazon.com/kms/latest/APIReference/API_CreateKey.html#API_CreateKey_RequestSyntax
export const CREATE_KEY_SPEC = 'RSA_2048';

export const RESOURCES_PER_PAGE = 10;

export const USER_AGENT = `${name}/${version}`;
