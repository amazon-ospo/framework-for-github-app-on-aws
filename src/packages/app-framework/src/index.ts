export * from './credential-manager/index';
