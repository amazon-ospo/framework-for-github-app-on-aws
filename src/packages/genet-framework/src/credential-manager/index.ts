import { RemovalPolicy, NestedStack, Tags } from 'aws-cdk-lib';
import { AttributeType, Table, BillingMode } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
export interface CredentialManagerProps {}

/**
 * Nested stack for storing GitHub App installation targets and pre-approved lists.
 */
export class CredentialManager extends NestedStack {
  readonly appTable: Table;
  readonly installationTable: Table;
  constructor(scope: Construct, id: string, props?: CredentialManagerProps) {
    super(scope, id, props);
    Tags.of(this).add('GenetComponent', 'CredentialManager');
    // Table for storing GitHub App IDs and their corresponding private key ARNs that stored in AWS KMS.
    this.appTable = new Table(this, 'AppTable', {
      partitionKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.RETAIN,
      pointInTimeRecovery: true,
    });
    Tags.of(this.appTable).add('CredentialManager', 'AppTable');
    // Table for tracking GitHub App installations.
    // Stores `node_id`, `installation_id`, `app_id`
    // and last updated timestamp.
    this.installationTable = new Table(this, 'AppInstallationTable', {
      partitionKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'NodeId',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      pointInTimeRecovery: true,
    });
    Tags.of(this.installationTable).add(
      'CredentialManager',
      'AppInstallationTable',
    );
    // Global secondary index for looking up installations by Node ID.
    this.installationTable.addGlobalSecondaryIndex({
      indexName: 'NodeID',
      partitionKey: {
        name: 'NodeId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
    });
    // Global secondary index for looking up installations by installations ID.
    this.installationTable.addGlobalSecondaryIndex({
      indexName: 'InstallationID',
      partitionKey: {
        name: 'InstallationId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
    });
  }
}
