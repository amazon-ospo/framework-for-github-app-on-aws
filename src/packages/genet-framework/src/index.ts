export * from './credential-manager/index';
export * from './installation-manager/index';
