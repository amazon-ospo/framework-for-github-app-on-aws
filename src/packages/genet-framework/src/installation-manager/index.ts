import { NestedStack } from 'aws-cdk-lib';
import { AttributeType, Table, BillingMode } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
export interface InstallationManagerProps {}
/**
 * Nested stack for managing GitHub App installation targets and pre-approved lists.
 */
export class InstallationManager extends NestedStack {
  readonly targetTable: Table;
  readonly installationTable: Table;
  constructor(scope: Construct, id: string, props: InstallationManagerProps) {
    super(scope, id, props);
    // Table for storing the pre-approval target list GitHub App installation.
    this.targetTable = new Table(this, 'PreApprovalTargetTable', {
      partitionKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      pointInTimeRecoverySpecification: { pointInTimeRecoveryEnabled: true },
    });
    // Global secondary index to query targets by Node ID.
    this.targetTable.addGlobalSecondaryIndex({
      indexName: 'NodeID',
      partitionKey: {
        name: 'NodeId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
    });
    // Table for tracking GitHub App installations.
    // Stores `node_id`, `installation_id`, `app_id`
    // and last updated timestamp.
    this.installationTable = new Table(this, 'AppInstallationTable', {
      partitionKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      pointInTimeRecoverySpecification: { pointInTimeRecoveryEnabled: true },
    });
    // Global secondary index for looking up installations by Node ID.
    this.installationTable.addGlobalSecondaryIndex({
      indexName: 'NodeID',
      partitionKey: {
        name: 'NodeId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
    });
    // Global secondary index for looking up installations by installations ID.
    this.installationTable.addGlobalSecondaryIndex({
      indexName: 'InstallationID',
      partitionKey: {
        name: 'InstallationId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'AppId',
        type: AttributeType.STRING,
      },
    });
  }
}
