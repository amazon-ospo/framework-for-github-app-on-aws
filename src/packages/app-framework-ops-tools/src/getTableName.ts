import {
  ResourceGroupsTaggingAPIClient,
  GetResourcesCommand,
  GetResourcesCommandOutput,
} from '@aws-sdk/client-resource-groups-tagging-api';
import { TagName, RESOURCES_PER_PAGE, USER_AGENT } from './constants';

const taggingClient = new ResourceGroupsTaggingAPIClient({
  customUserAgent: USER_AGENT,
});

/**
 * Fetches DynamoDB tables tagged with FRAMEWORK_FOR_GITHUB_APP_ON_AWS_MANAGED and CREDENTIAL_MANAGER
 * and APP_TABLE
 * @returns Array of table names that match the tag criteria
 */
export const listTablesByTags = async (): Promise<string[]> => {
  let allTables: string[] = [];
  let paginationToken: string | undefined;
  do {
    const command = new GetResourcesCommand({
      ResourceTypeFilters: ['dynamodb:table'],
      TagFilters: [
        {
          Key: TagName.FRAMEWORK_FOR_GITHUB_APP_ON_AWS_MANAGED,
          Values: [TagName.CREDENTIAL_MANAGER],
        },
        { Key: TagName.CREDENTIAL_MANAGER, Values: [TagName.APP_TABLE] },
      ],
      PaginationToken: paginationToken,
      ResourcesPerPage: RESOURCES_PER_PAGE,
    });

    try {
      const response: GetResourcesCommandOutput =
        await taggingClient.send(command);
      if (!!response && !!response.ResourceTagMappingList) {
        const tableNames = response.ResourceTagMappingList.map(
          (resource) => resource.ResourceARN?.split('/').pop()!,
        );
        allTables.push(...tableNames);
      }
      paginationToken = response.PaginationToken;
    } catch (error) {
      console.error('Error fetching tables:', error);
      throw error;
    }
  } while (paginationToken);

  return allTables;
};

export type DisplayDynamoDBTables = ({
  listTables,
}: {
  listTables?: () => Promise<string[]>;
}) => Promise<void>;

/**
 * Displays a list of DynamoDB tables that match Tags
 * with FrameworkForGitHubAppOnAwsManaged, CredentialManager and AppTable.
 * Lists tables in a numbered format and shows the total count.
 *
 * ---
 * dependency injection parameters:
 *
 * @param listTables Function that fetches the dynamoDB tables tagged with
 * FrameworkForGitHubAppOnAwsManaged, CredentialManager and AppTable
 *
 * @example
 * // Output:
 * // Available tables:
 * // 1. table1
 * // 2. table2
 * // Total tables found: 2
 */
export const displayDynamoDBTables: DisplayDynamoDBTables = async ({
  listTables = listTablesByTags,
}) => {
  try {
    const tables = await listTables();
    if (tables.length === 0) {
      throw new Error(
        'No tables found with the FRAMEWORK_FOR_GITHUB_APP_ON_AWS_MANAGED-APP_TABLE tag',
      );
    }
    console.log('\nAvailable tables:');
    tables.forEach((name, index) => {
      console.log(`${index + 1}. ${name}`);
    });
    console.log(`\nTotal tables found: ${tables.length}\n`);
  } catch (error) {
    throw error;
  }
};

if (require.main === module) {
  displayDynamoDBTables({}).catch((error) => {
    console.error('Unexpected error:', error);
    process.exit(1);
  });
}
