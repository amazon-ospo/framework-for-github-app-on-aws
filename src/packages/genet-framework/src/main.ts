import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { CredentialManager } from './credential-manager';
import { InstallationManager } from './installation-manager';

// CDK App entry for Genet acceptance test.
export class Genet extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    new CredentialManager(this, 'CredentialManager', {});
    new InstallationManager(this, 'InstallationManager', {});
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new Genet(app, 'genet-framework', { env: devEnv });

app.synth();