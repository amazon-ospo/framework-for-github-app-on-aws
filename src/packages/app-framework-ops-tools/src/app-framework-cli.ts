#!/usr/bin/env node

import { Command } from 'commander';
import { version } from '../package.json';
import { displayDynamoDBTables } from './getTableName';
import { importPrivateKey } from './importPrivateKey';

const program = new Command();
/**
 * CLI tool for Importing GitHub App private key into AWS KMS
 * Main command - 'app-framework'
 */
program
  .name('app-framework')
  .description(
    'CLI tool to get name of the App table with FrameworkForGitHubAppOnAwsManaged tag and to import GitHub App private key into AWS KMS',
  )
  .version(version);
// subcommand - getTableName
program
  .command('getTableName')
  .description('Displays App tables with FrameworkForGitHubAppOnAwsManaged tag')
  .action(async () => {
    try {
      await displayDynamoDBTables({});
    } catch (error) {
      console.error('Error:', error);
      process.exit(1);
    }
  });
// subcommand - importPrivateKey
program
  .command('importPrivateKey')
  .description('Import GitHub App private key into AWS KMS')
  .argument('<pemFilePath>', 'Path to the private key PEM file')
  .argument('<appId>', 'GitHub App ID')
  .argument('<tableName>', 'Table name to store the AppId and Key ARN')
  .addHelpText(
    'after',
    `
    Example:
      $ app-framework importPrivateKey private-key.pem 123456 my-table-name
  `,
  )
  .action(
    async (pemFilePath: string, appIdAsString: string, tableName: string) => {
      try {
        const appId = Number(appIdAsString);
        if (isNaN(appId)) {
          console.error('Error: GitHub AppId must be a valid number');
          process.exit(1);
        }

        await importPrivateKey({ pemFilePath, appId, tableName });
      } catch (error) {
        console.error('Error:', error);
        process.exit(1);
      }
    },
  );
program.parse(process.argv);
