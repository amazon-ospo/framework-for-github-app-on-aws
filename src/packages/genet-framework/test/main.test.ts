import { Hello } from '../src';
/**
 * TODO: After we start adding things in main.test.ts, delete this.
 */
test('hello', () => {
  expect(new Hello().sayHello()).toBe('hello, world!');
});
