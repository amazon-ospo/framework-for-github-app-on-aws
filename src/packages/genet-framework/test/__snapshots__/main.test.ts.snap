// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CredentialManagerNestedStackCredentialManagerNestedStackResource341E8524": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region",
              },
              ".",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/dfe1e02adf2e9c28f3f1412d597ffd1972ac0571ab9392ea95f3bf5f3950ccdf.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "InstallationManagerNestedStackInstallationManagerNestedStackResourceBB60C6B1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region",
              },
              ".",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/36735340e49e5e322a367be3be3d174093f55af585324996aebb499fd62104a5.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
